homeassistant/components/fireservicerota/__init__.py

```diff
@@ -0,0 +1,182 @@
+        self._hass = hass
+        self._entry = entry
+
+        self._url = entry.data[CONF_URL]
+        self._tokens = entry.data[CONF_TOKEN]
+
+        self.availability_data = None
+        self.incident_data = None
+        self.incident_id = None
+        self.response_data = None
+
+        self.fsr_avail = FireServiceRota(
+            base_url=f"https://{self._url}", token_info=self._tokens
+        )
+
+        self.fsr_incidents = FireServiceRotaIncidents(on_incident=self.on_incident)
+
+        self.start_ws_listener()
+
+    def construct_url(self) -> str:
+        """Return url with latest config values."""
+        return WSS_BWRURL.format(
+            self._entry.data[CONF_URL], self._entry.data[CONF_TOKEN]["access_token"]
+        )
+
+    def on_incident(self, data) -> None:
+        """Update the current data."""
+        _LOGGER.debug("Got data from websocket listener: %s", data)
+        self.incident_data = data
+
+        async_dispatcher_send(self._hass, f"{DOMAIN}_{self._entry.entry_id}_update")
+
+    def start_ws_listener(self) -> None:
+        """Start the websocket listener."""
+        _LOGGER.debug("Starting incidents listener")
+        self.fsr_incidents.start(self.construct_url())
+
+    def stop_ws_listener(self) -> None:
+        """Stop the websocket listener."""
+        _LOGGER.debug("Stopping incidents listener")
+        self.fsr_incidents.stop()
+
+    async def update_call(self, func, *args):
+        """Perform update call and return data."""
+        try:
+            return await self._hass.async_add_executor_job(func, *args)
+        except (ExpiredTokenError, InvalidTokenError):
+            if await self.async_refresh_tokens():
+                return await self._hass.async_add_executor_job(func, *args)
+
+    @Throttle(MIN_TIME_BETWEEN_UPDATES)
+    async def async_update(self) -> None:
+        """Get the latest availability data."""
+        _LOGGER.debug("Updating availability data")
+
+        self.availability_data = await self.update_call(
+            self.fsr_avail.get_availability, str(self._hass.config.time_zone)
+        )
+
+    async def async_response_update(self) -> None:
+        """Get the latest incident response data."""
+        if self.incident_data is None:
+            return
+
+        self.incident_id = self.incident_data.get("id")
+        _LOGGER.debug("Updating incident response data for id: %s", self.incident_id)
+
+        self.response_data = await self.update_call(
+            self.fsr_avail.get_incident_response, self.incident_id
+        )
+
+    async def async_set_response(self, incident_id, value) -> None:
+        """Set incident response status."""
+        _LOGGER.debug("Setting incident response status")
+
+        await self.update_call(self.fsr_avail.set_incident_response, incident_id, value)
+
+    async def async_refresh_tokens(self) -> bool:
+        """Refresh tokens and update config entry."""
+        self.stop_ws_listener()
```

Comment:
Should the WS listener only be stopped if refresh was successful?

Reply:
When there are token issues, Websocket Client is stopped asap, otherwise a bunch of errors is created, and is started again when tokens are solved (the last events are then picked up, so nothing is missed, only a small delay is created)